File: CPP/20.cpp
class Solution {
public:
    bool isValid(string s) {
        // Stack will be the best choince
        // } and ) must match the stack top.
        vector<char> stk;
        for (char ch : s){
            cout << ch << " ";
            if (ch == '(' | ch == '{' | ch == '['){
                stk.push_back(ch);
            } else {
                if (stk.empty()) return false;
                if (ch == ']' & stk.back() == '[') stk.pop_back();
                else if (ch == ')' & stk.back() == '(') stk.pop_back();
                else if (ch == '}' & stk.back() == '{') stk.pop_back();
                else return false;
            }
        }
        cout << endl;
        return (stk.empty()) ? true:false;
    }
};

File: Python/20.py
class Solution:
    def isValid(self, s: str) -> bool:
        dic = {'(' : ')', '{' : '}', '[' : ']'}
        # first element is the key
        stk = []
        for ch in s:
            if ch in dic: # equal to == or == ...
                stk.append(ch)
            elif len(stk) == 0 or dic[stk.pop()] is not ch:
                return False
        return len(stk) == 0
